grails-app/services/org/pih/warehouse/importer/PersonImportDataService.groovy

```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright (c) 2012 Partners In Health.  All rights reserved.
+ * The use and distribution terms for this software are covered by the
+ * Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
+ * which can be found in the file epl-v10.html at the root of this distribution.
+ * By using this software in any fashion, you are agreeing to be bound by
+ * the terms of this license.
+ * You must not remove this notice, or any other, from this software.
+ **/
+package org.pih.warehouse.importer
+
+import grails.gorm.transactions.Transactional
+import org.pih.warehouse.core.Person
+import org.pih.warehouse.data.PersonService
+import org.springframework.validation.BeanPropertyBindingResult
+
+@Transactional
+class PersonImportDataService implements ImportDataService {
+    PersonService personService
+
+    void validateData(ImportDataCommand command) {
+        command.data.eachWithIndex { params, index ->
+            Person person = personService.createOrUpdatePerson(params)
+            if (!person.validate()) {
+                person.errors.each { BeanPropertyBindingResult error ->
+                    command.errors.reject("Row ${index + 1} name = ${person.name}: ${error.getFieldError()}")
+                }
+            }
+        }
+    }
+
+    void importData(ImportDataCommand command) {
+        command.data.eachWithIndex { params, index ->
+            Person person = personService.createOrUpdatePerson(params)
+            if (person.validate()) {
```

Comment:
this is what I meant when saying that `create` method should also proceed with the save - I think it should be `create` method's responsibility to persist the data.
Additional note - to follow "good practices" we should call the `.save()` only if we create a brand new object (or some associations inside it is new), because calling `.save()` for a persistent entity is unnecessary, since Hibernate will persist it at the transaction commit.

Reply:
No no no. 

createOrUpdate* should not persist. Ever. They are used for populating an object that can be validated and persisted by caller. This is because sometimes we call createOrUpdate in the validateData() method becuase we need to construct the data object before it is validated and we don't want the validateData method to persist the object. The validateData and importData are completely separate steps in the import process. 

Also, we don't want createOrUpdate* to persist for performance reasons. Read this https://www.naleid.com/2009/10/01/batch-import-performance-with-grails-and-mysql.html

So if createOrUpdate is confusing, then we need to rename to something that makes it clear that we are constructing the object, not persisting the object.

Reply:
The Person handler is a simple encapsulation of why we shouldn't persist in the createOrUpdate method 
```
    Boolean validateData(ImportDataCommand command) {
        command.data.eachWithIndex { params, index ->
            Person person = createOrUpdatePerson(params)
            if (!person.validate()) {
                person.errors.each { BeanPropertyBindingResult error ->
                    command.errors.reject("Row ${index + 1} name = ${person.name}: ${error.getFieldError()}")
                }
            }
        }
    }

    void importData(ImportDataCommand command) {
        command.data.eachWithIndex { params, index ->
            Person person = createOrUpdatePerson(params)
            if (person.validate()) {
                person.save(failOnError: true)
            }
        }
    }
```
and why we don't need a separate command class.
```
    Person createOrUpdatePerson(Map params) {
        Person person = Person.findByFirstNameAndLastName(params.firstName, params.lastName)
        if (!person) {
            person = new Person(params)
        } else {
            person.properties = params
        }
        return person
    }
```

Reply:
@jmiranda as for performance side, I quite don't understand what current code does in order to improve the performance? 
At the moment we still call the save inside the loop:
```groovy

    void importData(ImportDataCommand command) {
        command.data.eachWithIndex { params, index ->
            Person person = createOrUpdatePerson(params)
            if (person.validate()) {
                person.save(failOnError: true)
            }
        }
    }
```
so wouldn't calling the save inside the `createOrUpdatePerson` actuallly be just moving the `if (person.validate()) {...}` part there? which in fact doesn't change anything?

I know that in Spring there is `saveAll()` method which might improve performance for some batch operations. As I can see, in Grails we also have it, but from the domain level, e.g.`User.saveAll()` - maybe it could be considered somewhere for an experiment :man_shrugging:

Reply:
Tech Huddle: Add tech debt to look into performance difference between calling save on individual objects or saveAll on a collection of objects.